crypto/property/property_parse.c

```diff
@@ -588,15 +588,38 @@ static void put_char(char ch, char **buf, size_t *remain, size_t *needed)
 
 static void put_str(const char *str, char **buf, size_t *remain, size_t *needed)
 {
-    size_t olen, len;
+    size_t olen, len, i;
+    char quote = '\0';
+    int quotes;
 
     len = olen = strlen(str);
     *needed += len;
 
-    if (*remain == 0)
+    /*
+     * Check to see if we need quotes or not.
+     * Characters that are legal in a PropertyName don't need quoting.
+     * We simply assume all others require quotes.
+     */
+    for (i = 0; i < len; i++)
+        if (!ossl_isalnum(str[i]) && str[i] != '.' && str[i] != '_') {
+            /* Default to single quotes ... */
+            if (quote == '\0')
+                quote = '\'';
+            /* ... but use double quotes if a single is present */
+            if (str[i] == '\'')
+                quote = '"';
+        }
+
+    quotes = quote != '\0';
+    if (*remain == 0) {
+        *needed += 2 * quotes;
         return;
+    }
 
-    if (*remain < len + 1)
+    if (quotes)
+        put_char(quote, buf, remain, needed);
+
+    if (*remain < len + 1 + quotes)
         len = *remain - 1;
```

Comment:
Should there be `len = *remain - 1 - quotes;`? Also is truncating the string right here?

Reply:
I don't think the closing quote is required.  We're truncating the value, so it's never going to parse back to a query correctly.  Omitting the closing quote is thus harmless.

The truncation already existed.  I can't think of a better thing to do when space runs out.

Reply:
By omitting the closing quote, the property parser will definitely produce an error.
This would not be the case if it were included -- it likely would parse okay but do the wrong thing.

Reply:
Hmm... but what about unquoted strings - they are truncated silently.

Reply:
Yep, that was the existing behaviour.  @mattcaswell introduced this function.

What should it do on out of space?

Not that out of space is possible in any current usage since a buffer of the correct size is allocated.

Reply:
Could you make it returning int and simply return a failure?

Reply:
It needs to return the length, so callers can allocate appropriate space.  We can't stop early with an error.
It might be possible to cache the error condition but continue, ugly though.

Reply:
Not sure you have to make it an error condition.  You could model it along the lines of what snprintf() does, and have it return _the size it would take_ (minus the ending NUL byte) unconditionally.  That leaves it to the caller to check if that number is higher or equal to the allocated space, and deal with it.

Reply:
We do return the size it would take (including the terminating byte).

The question is: _what should we do with the buffer when the passed size isn't large enough?_

@mattcaswell, this was your decision originally, please let us know your thoughts about why you chose to truncate rather than something else???

I really think that this should be a discussion for a different PR/issue.  The code here does **exactly** what the existing code does for everything except quoted strings (which the original code screwed up).

Reply:
Ideally, this should be rewritten using WPACKET to centralise the buffering etc.

We only ever call this with a NULL pointer or a malloced buffer, so it would simplify the callers by having WPACKET allocate the space.

Still, I don't think that's something to do here.

Reply:
Sorry, somehow missed the pings on this. Truncation is the documented behaviour for the `ossl_property_list_to_string` function:

https://github.com/openssl/openssl/blob/master/doc/internal/man3/ossl_global_properties_no_mirrored.pod

It is the caller's responsibility to ensure that the buffer is big enough - which it can do by calling `ossl_property_list_to_string` initially with an empty buffer, and allocating the number of bytes returned. So, if the caller is behaving properly, then truncation will either never occur or, if it does, presumably the caller doesn't care about possible truncation. All current callers of this do it correctly.

Whether truncation is the correct thing to do or not, I'm not sure. I don't recall my thinking at the time. But I don't think it is this PRs responsibility to change that. It should continue to truncate IMO.

Reply:
Thanks @mattcaswell.

I still think that WPACKET would be the better approach.